sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,262 @@
+# -*- coding: utf-8 -*-
+"""K-medoids clustering"""
+
+# Authors: Timo Erkkilä <timo.erkkila@gmail.com>
+#          Antti Lehmussola <antti.lehmussola@gmail.com>
+# License: BSD 3 clause
+
+import numpy as np
+import warnings
+
+from ..base import BaseEstimator, ClusterMixin, TransformerMixin
+from ..metrics.pairwise import PAIRWISE_DISTANCE_FUNCTIONS
+from ..utils import check_array
+
+# With Python 2.x ValueError needs to be imported,
+# otherwise that's not needed
+try:
+    from exceptions import ValueError
+except ImportError:
+    pass
+
+
+class KMedoids(BaseEstimator, ClusterMixin, TransformerMixin):
+    """
+    k-medoids class.
+
+    Parameters
+    ----------
+    n_clusters: int, optional, default: 8
+      How many medoids. Must be positive.
+
+    distance_metric: string, optional, default: 'euclidean'
+      What distance metric to use.
+
+    clustering: {'pam'}, optional, default: 'pam'
+      What clustering mode to use.
+
+    init: {'random', 'heuristic'}, optional, default: 'heuristic'
+      Specify medoid initialization.
+    """
+
+    # Supported clustering methods
+    CLUSTERING_METHODS = ['pam']
+
+    # Supported initialization methods
+    INIT_METHODS = ['random', 'heuristic']
+
+    def __init__(self, n_clusters=8, distance_metric='euclidean',
+                 clustering_method='pam', init='heuristic'):
+
+        # Check n_clusters
```

Comment:
Please do not do any parameter validation in `__init__`, but in `fit` instead.
More details [here](http://scikit-learn.org/stable/developers/index.html#parameters-and-init)